### PR TITLE
Fakify script object inputs and attributes for non-strict export

### PR DESCRIPTION
**Pull request by ydwu4**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124978
* __->__ #124239

This PR fakify ScriptObject inputs and attributes in export non-strict mode by default. 

The basic idea is to `only fakify the script object during tracing (i.e. aot_export)`. After we get the traced graph module, eagerly executing, serializing, or running more passes will use the real script objects. This is essentially treating the script object as constant tensor.

Concretely, we 
1. fakify all the script object inputs, and module attributes (gathered by constant_attrs). 
2. patch the module's attributes with fakified script object 
3. right after aot_export, remove the patching (to avoid changing the original module) then modify the exported graph module's attribute to real script object.

### CODE REVIEW

File: torch/_export/non_strict_utils.py [hunk 1]
```diff
@@ -301,3 +307,115 @@ def _is_dynamic_shape_leaf(x):
             range_constraints[symbol] = shape_env.var_to_range[symbol]
 
     return range_constraints
+
+
+def _gather_constant_attrs(m: torch.nn.Module) -> ConstantAttrMap:
```
Comment (ydwu4) [inline torch/_export/non_strict_utils.py:312]: code movement only for `_gather_constant_attrs`

File: torch/_export/passes/lift_constants_pass.py [hunk 2]
```diff
@@ -229,24 +243,36 @@ def lift_constants_pass(
 
 def rewrite_script_object_meta(
     gm: torch.fx.GraphModule,
-) -> Dict[str, Union[torch.Tensor, torch.ScriptObject]]:
+) -> Dict[str, Union[torch.Tensor, FakeScriptObject,],]:
     """When tracing, we produce a graph with an actual ScriptObject in the
```
Comment (ydwu4) [inline torch/_export/passes/lift_constants_pass.py:247]: modify comments

File: torch/export/_trace.py [hunk 3]
```diff
@@ -988,16 +972,32 @@ def forward(self, *args, **kwargs):
...
-            ep_non_strict = _export_non_strict(
-                mod,
-                fake_args,
-                fake_kwargs,
-                fake_params_buffers,
-                constant_attrs,
-                pre_dispatch=pre_dispatch,
-                transform=_tuplify_outputs,
-            )
+            with _fakify_script_objects(mod, fake_args, fake_kwargs, fake_mode) as (
```
Comment (ydwu4) [inline torch/export/_trace.py:977]: Should probably move `_fakify_script_objects` into `_export_non_strict` since strict mode would also benefit from this. and also move the patching-back logic of ep_non_strict.constants into _export_non_strict

File: torch/export/_trace.py [hunk 4]
```diff
@@ -1156,6 +1156,7 @@ def forward(self, *args, **kwargs):
     _normalize_nn_module_stack(gm_torch_level, type(mod))
 
     # NOTE: graph module expects only positional args
+    constant_attrs = _gather_constant_attrs(mod)
```
Comment (ydwu4) [inline torch/export/_trace.py:1159]: Remove this one, this _export_nont_strict will deal with this after the movement mentioned in previous comment is done.

File: torch/fx/_symbolic_trace.py [hunk 5]
```diff
@@ -729,11 +729,17 @@ def trace(
...
+                    torch.Tensor,
+                    torch.ScriptObject,
+                    FakeScriptObject
+                ], str
+            ] = {}
 
             def collect_tensor_attrs(m: torch.nn.Module, prefix_atoms: List[str]):
                 for k, v in m.__dict__.items():
-                    if isinstance(v, (torch.Tensor, ScriptObject)):
+                    if isinstance(v, (torch.Tensor, FakeScriptObject)):
```
Comment (ydwu4) [inline torch/fx/_symbolic_trace.py:742]: add back ScriptObject type for bc reasons.

File: test/export/test_passes.py [hunk 6]
```diff
@@ -87,6 +94,55 @@ def _get_output_names(gm: torch.fx.GraphModule) -> List[str]:
...
+    class Simple(torch.nn.Module):
+        def __init__(self):
+            super().__init__()
+            self.attr = torch.classes._TorchScriptTesting._Foo(10, 20)
+
+    class SimpleWithAttrInContainer(torch.nn.Module):
+        def __init__(self):
+            super().__init__()
+            self.attr = torch.classes._TorchScriptTesting._Foo(10, 20)
+            self.pytree_attr2 = [
```
Comment (ydwu4) [inline test/export/test_passes.py:109]: currently, _gather_constant_attrs doesn't recursively gather the constants inside pytree container attributes. Need to fix it in the future.

File: torch/testing/_internal/torchbind_impls.py [hunk 7]
```diff
@@ -30,3 +32,91 @@ def fake_queue_push(tq, x):
...
+        raise unittest.SkipTest("non-portable load_library call used in test")
+    else:
+        lib_file_path = find_library_location("libtorchbind_test.so")
+        if IS_WINDOWS:
+            lib_file_path = find_library_location("torchbind_test.dll")
+        torch.ops.load_library(str(lib_file_path))
+
+
+@contextlib.contextmanager
+def _register_py_impl_temporially(op_overload, key, fn):
```
Comment (zou3519) [inline torch/testing/_internal/torchbind_impls.py:116]: nit: temporarily

File: torch/export/_trace.py (hunk 3 above)
Comment (zou3519) [reply torch/export/_trace.py:977]: Sounds legit, are you doing this in this PR or another one?

File: torch/export/_trace.py [hunk 8]
```diff
@@ -988,16 +972,32 @@ def forward(self, *args, **kwargs):
...
-                pre_dispatch=pre_dispatch,
-                transform=_tuplify_outputs,
-            )
+            with _fakify_script_objects(mod, fake_args, fake_kwargs, fake_mode) as (
+                patched_mod,
+                new_fake_args,
+                new_fake_kwargs,
+                new_fake_constant_attrs,
+                map_fake_to_real,
+            ):
```
Comment (zou3519) [inline torch/export/_trace.py:983]: I don't really get why `_fakify_script_objects` returns new_fake_args and new_fake_kwargs -- are we saying that ScriptObjects can show up in both args and as constants and this is handling both?

File: torch/_ops.py [hunk 9]
```diff
@@ -870,8 +871,13 @@ def _dispatch_in_python(self, args, kwargs, fallthrough_keys):
...
             raise RuntimeError(
-                f"Cannot handle FakeScriptObject with python dispatcher with dispatch key {handler}."
-                f"Please implement it by annotating a python callable with py_impl({handler})."
+                f"Torchbind op {self} received a FakeScriptObject input when dispatching {handler}."
+                f" but no python implementation is found."
+                f" This error can happen when you export or compile the model."
+                f" To fix it, you can register a python impl by torch.ops.{self}.py_impl(torch._C.{handler})(...)."
+                f" Note this error can still happpen even if a C++ implementation "
+                f" has been registered. That's because FakeScriptObject purely lives in python and cannot work "
+                f" with a C++ implementation."
```
Comment (zou3519) [inline torch/_ops.py:880]: nit: ask them to file an issue, they're probably not going to be able to get this to work without our help

File: torch/_library/fake_class_registry.py [hunk 10]
```diff
@@ -10,9 +10,12 @@
 
 
 class FakeScriptObject:
-    def __init__(self, wrapped_obj):
+    def __init__(self, wrapped_obj: Any, qualified_name: str):
```
[the remark this answers is not in the document]
Comment (zou3519) [reply torch/_library/fake_class_registry.py:13]: What is the qualified name?

File: torch/_export/passes/lift_constants_pass.py [hunk 11]
```diff
@@ -15,33 +17,35 @@
...
 
 class ConstantAttrMap(collections.abc.MutableMapping):
-    """A mapping class that understands how to use module constants (tensors and
-    ScriptObjects) as keys. We store tensors normally, but ScriptObjects are
-    stored by hash, because different torch.ScriptObjects can point to the same
-    underlying value (but we guarantee that they will `hash()` to the same value
+    """A mapping class that understands how to use module constants (tensors,
+    ScriptObjects, FakeScriptObjects) as keys. We store tensors and FakeScriptObjects normally,
+    but ScriptObjects are stored by hash, because different torch.ScriptObjects can point to
+    the same underlying value (but we guarantee that they will `hash()` to the same value
```
Comment (zou3519) [inline torch/_export/passes/lift_constants_pass.py:23]: I'm a bit lost, do we still have ScriptObjects here? I thought that only FakeScriptObjects should show up during tracing.

**Review by zou3519 — COMMENTED** [review]
Took a pass through; the approach sounds good to me. Left some comments. In general I was confused about where we only have FakeScriptObject and where we can sometimes have both FakeScriptObject and ScriptObject

File: torch/export/_trace.py (hunk 8 above)
Comment (zou3519) [reply torch/export/_trace.py:983]: Yes, it looks like this is the case

File: torch/export/_trace.py (hunk 3 above)
Comment (ydwu4) [reply torch/export/_trace.py:977]: I'll do it in this one.

Comment (ydwu4) [reply torch/export/_trace.py:977]: I changed my mind. We probably should wait for dynamo support landed first.

File: torch/testing/_internal/torchbind_impls.py [hunk 12]
```diff
@@ -30,3 +32,99 @@ def fake_queue_push(tq, x):
     @register_if_not("_TorchScriptTesting::queue_size")
     def fake_queue_size(tq):
         return tq.size()
+
```
Comment (zou3519) [inline torch/testing/_internal/torchbind_impls.py:35]: cleaner thing to do: make sure register_fake_operators gets run only once instead of if-then-else everywhere below

File: torch/_library/fake_class_registry.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (zou3519) [reply torch/_library/fake_class_registry.py:13]: yeah just rename this to "class_name" or something

File: test/export/test_passes.py (hunk 6 above)
Comment (zou3519) [reply test/export/test_passes.py:108]: Please add a test that the pytree thing doesn't work (and label it with an expectedFailure or something). Maybe also an issue

**Review by zou3519 — APPROVED** [review]

File: torch/_export/non_strict_utils.py [hunk 13]
```diff
@@ -319,3 +324,115 @@ def make_constraints(
...
+    assert not pytree.tree_any(
+        lambda obj: isinstance(obj, FakeScriptObject), (args, kwargs)
+    ), "args and kwargs shouldn't contain any FakeScriptObject."
+
+    patched_attr = {}
+    fake_constant_attrs = ConstantAttrMap()
+    fake_to_real = {}
+
+    def _fakify_obj(obj):
+        fake_obj = torch._library.fake_class_registry.to_fake_obj(fake_mode, obj)
```
Comment (ydwu4) [inline torch/_export/non_strict_utils.py:400]: It's too restrictive right now. To maintain BC and allow gradual migration, we should allow real script object to show up for now and address them one-by-one.

File: torch/_export/non_strict_utils.py [hunk 14]
```diff
@@ -319,3 +324,115 @@ def make_constraints(
...
+                fake_constant_attrs[fake_script_obj] = fqn
+                patched_attr[fqn] = obj
+            else:
+                fake_constant_attrs[obj] = fqn
+
+        fake_args, fake_kwargs = pytree.tree_map_only(
+            torch.ScriptObject, _fakify_obj, (args, kwargs)
+        )
+        assert not any(
+            isinstance(obj, torch.ScriptObject) for obj in fake_constant_attrs.values()
```
Comment (ydwu4) [inline torch/_export/non_strict_utils.py:429]: Remove the assertions.

File: torch/_export/passes/lift_constants_pass.py [hunk 15]
```diff
@@ -229,24 +243,35 @@ def lift_constants_pass(
...
+) -> Dict[str, Union[torch.Tensor, FakeScriptObject],]:
+    """When tracing, we produce a graph with FakeScriptObject in the
+    meta["val"].
 
     For now, we rewrie meta["val"] to be a placeholder CustomObjArgument
     """
-    constants: Dict[str, Union[torch.Tensor, torch._C.ScriptObject]] = {}
+    constants: Dict[
+        str,
+        Union[
```
Comment (ydwu4) [inline torch/_export/passes/lift_constants_pass.py:254]: add torch.ScriptObject back.

File: torch/_export/passes/lift_constants_pass.py [hunk 16]
```diff
@@ -229,24 +243,35 @@ def lift_constants_pass(
...
-        ):
+        if "val" not in node.meta:
             continue
 
-        old_meta = node.meta["val"]
-        class_fqn = old_meta._type().qualified_name()  # type: ignore[attr-defined]
-        new_meta = CustomObjArgument(node.name, class_fqn)
-        constants[node.name] = old_meta
-        node.meta["val"] = new_meta
+        assert not isinstance(
```
Comment (ydwu4) [inline torch/_export/passes/lift_constants_pass.py:263]: Remove the assertion

File: torch/export/_trace.py [hunk 17]
```diff
@@ -494,6 +466,9 @@ def _export_non_strict(
     transform=lambda x: x,  # TODO(zhxchen17) Revisit if this is needed later.
     pre_dispatch=False,
 ):
+    assert not any(
```
Comment (ydwu4) [inline torch/export/_trace.py:469]: remove the assertion

File: torch/export/_trace.py [hunk 18]
```diff
@@ -626,7 +599,14 @@ def make_argument_spec(i, node) -> ArgumentSpec:
     )
 
     constants = rewrite_script_object_meta(gm)
-    constants.update(lift_constants_pass(gm, export_graph_signature, constant_attrs))
+    attr_constants = lift_constants_pass(gm, export_graph_signature, constant_attrs)
+    assert not any(
+        isinstance(obj, torch.ScriptObject) for obj in attr_constants.values()
```
Comment (ydwu4) [inline torch/export/_trace.py:604]: Remove the assertion

File: torch/export/_trace.py [hunk 19]
```diff
@@ -643,7 +623,13 @@ def make_argument_spec(i, node) -> ArgumentSpec:
     class _ExportedProgramNonStrict:
         gm: torch.fx.GraphModule
         sig: ExportGraphSignature
-        constants: Dict[str, Union[torch.Tensor, torch._C.ScriptObject]]
+        constants: Dict[
+            str,
+            Union[
+                torch.Tensor,
```
Comment (ydwu4) [inline torch/export/_trace.py:629]: add back torch.ScriptObject

File: torch/export/unflatten.py [hunk 20]
```diff
@@ -54,7 +55,16 @@ def _assign_attr(
         assert isinstance(from_obj, torch.Tensor)
         to_module.register_buffer(field, from_obj, persistent=persistent)
     elif attr_kind == _AttrKind.CONSTANT:
-        assert isinstance(from_obj, (torch.Tensor, torch.ScriptObject))
+        assert not isinstance(
```
Comment (ydwu4) [inline torch/export/unflatten.py:58]: Remove assertion
